Generate module rundown events

MichalStrehovsky opened the pull request:
This is sufficient to get module information in `dotnet-gcdump collect`. The tool itself will need tweaks because it only looks at IL paths right now, but this should be all that's necessary to build on top.

Cc @brianrob @noahfalk @dotnet/ilc-contrib

MichalStrehovsky committed:
Generate module rundown events

This is sufficient to get module information in `dotnet-gcdump collect`. The tool itself will need tweaks because it only looks at IL paths right now, but this should be all that's necessary to build on top.

MichalStrehovsky committed:
Put back GUID

MichalStrehovsky committed:
Fix linux

MichalStrehovsky committed:
More Linux fixing

MichalStrehovsky committed:
linux

elinor-fung commented on src/coreclr/nativeaot/Runtime/eventtrace.cpp:
```diff
@@ -302,8 +325,47 @@ void EtwCallback(
+/**************************************************************************************/
+/* Called when ETW is turned OFF on an existing process .Will be used by the controller for end rundown*/
+/**************************************************************************************/
+void ETW::EnumerationLog::EndRundown()
```
Should this also fire `DCEndInit` and `DCEndComplete`? I assume they are not needed for `dotnet-gcdump collect`, but they are part of the contract for how rundown works.

elinor-fung replied:
They are basically indicators for tools to know when the rundown actually starts/completes (and the tool should have all the data, so can stop collecting). As long as they are fired before/after the rest of the end rundown events, it should be good.

noahfalk reviewed (approved):
Thanks @MichalStrehovsky, this looks nice!

elinor-fung commented on src/coreclr/nativeaot/Runtime/eventtrace.cpp:
```diff
@@ -302,8 +325,47 @@ void EtwCallback(
+        PalGetPDBInfo(pModule, &nativeGuid, &dwAge, wszPath, ARRAY_SIZE(wszPath));
+
+        GUID zeroGuid = { 0 };
+        FireEtwModuleDCEnd_V2(
```
Is this information only coming through the `ModuleDCEnd` event? If so, I think it should also be in a `ModuleLoad` event that gets fired on start. If a tool enables collection at before startup / for the lifetime of a process, it shouldn't have to trigger a rundown to get additional information (since it shouldn't have 'missed' anything).

elinor-fung replied:
Maybe at the end of `InitializeRuntime` - after `InitializeModules`, before we call the user's managed code?

elinor-fung commented on src/coreclr/nativeaot/Runtime/eventtrace.cpp:
```diff
@@ -282,6 +302,9 @@ void EtwCallback(
         providerContext = MICROSOFT_WINDOWS_DOTNETRUNTIME_PROVIDER_DOTNET_Context;
+    } else if (context->RegistrationHandle == Microsoft_Windows_DotNETRuntimeRundownHandle) {
+        providerIndex = DotNETRuntimeRundown;
+        providerContext = MICROSOFT_WINDOWS_DOTNETRUNTIME_RUNDOWN_PROVIDER_DOTNET_Context;
```
Does this callback also need to check for `EndRundownKeyword` to trigger a rundown for ETW scenarios?

MichalStrehovsky committed:
wow

MichalStrehovsky committed:
Feedback

MichalStrehovsky commented on src/coreclr/nativeaot/Runtime/eventtrace.cpp:
```diff
@@ -292,18 +315,110 @@ void EtwCallback(
-    if (IsEnabled &&
-        (context->RegistrationHandle == Microsoft_Windows_DotNETRuntimePrivateHandle) &&
-        GCHeapUtilities::IsGCHeapInitialized())
+    if (IsEnabled)
```
The `IsEnabled` here is actually `ControlCode` and these checks look a bit suspicious. I didn't touch that part, but I can if you agree this looks wrong.

elinor-fung replied:
I think this is functionally okay, but definitely looks suspicious and confusing. For rundown, triggering the end (or start) is an explicit action via control codes of either `EVENT_CONTROL_CODE_ENABLE_PROVIDER` or `EVENT_CONTROL_CODE_CAPTURE_STATE` with the end rundown keyword enabled. I think this is working only because those values are 1 and 2 respectively and `EVENT_CONTROL_CODE_DISABLE_PROVIDER` is 0. I'm not sure about the private runtime events, but I suspect it is a similar idea - just via the private provider itself since there is no private rundown provider.

Maybe:
- rename `IsEnabled` to `ControlCode`
- make this an explicit check against `EVENT_CONTROL_CODE_ENABLE_PROVIDER` or `EVENT_CONTROL_CODE_CAPTURE_STATE`
- add a comment about how these events are fired based on this explicit action/trigger

MichalStrehovsky committed:
Maybe it works?

jkotas commented on src/coreclr/nativeaot/Runtime/windows/PalCommon.cpp:
```diff
@@ -34,6 +34,184 @@ void PalGetModuleBounds(HANDLE hOsHandle, _Out_ uint8_t ** ppLowerBound, _Out_ u
+
+        DWORD rvaOfRawData = rgDebugEntries[i].AddressOfRawData;
+        ULONG cbDebugData = rgDebugEntries[i].SizeOfData;
+        if (cbDebugData < size_t(&((CV_INFO_PDB70*)0)->magic) + sizeof(((CV_INFO_PDB70*)0)->magic))
```
```suggestion
        if (cbDebugData < offsetof(CV_INFO_PDB70, magic) + sizeof(((CV_INFO_PDB70*)0)->magic))
```

jkotas commented on src/coreclr/nativeaot/Runtime/windows/PalCommon.cpp:
```diff
@@ -34,6 +34,184 @@ void PalGetModuleBounds(HANDLE hOsHandle, _Out_ uint8_t ** ppLowerBound, _Out_ u
+        DWORD          magic; 
+        GUID           signature;       // unique identifier 
+        DWORD          age;             // an always-incrementing value 
+        _Field_z_ char  path[MAX_PATH];  // zero terminated string with the name of the PDB file 
```
This is a stale or mangled definition of this structure. The up-to-date definition looks like this (you can check this in the MS internal OS repo) - notice that the file path is an open array:

```
struct CV_INFO_PDB70 
{
	DWORD      CvSignature; 
	GUID       Signature;       // unique identifier 
	DWORD      Age;             // an always-incrementing value 
	BYTE       PdbFileName[1];  // zero terminated string with the name of the PDB file 
};
```

jkotas replied:
```suggestion
        BYTE           path[MAX_PATH];  // zero terminated string with the name of the PDB file 
```

jkotas commented on src/coreclr/nativeaot/Runtime/windows/PalCommon.cpp:
```diff
@@ -34,6 +34,184 @@ void PalGetModuleBounds(HANDLE hOsHandle, _Out_ uint8_t ** ppLowerBound, _Out_ u
+        // outright corrupt.
+
+        // Verify sane size of raw data
+        if (cbDebugData > sizeof(CV_INFO_PDB70))
```
This is bogus sanity check. It is going to reject long paths incorrectly.

jkotas commented on src/coreclr/nativeaot/Runtime/windows/PalCommon.cpp:
```diff
@@ -34,6 +34,184 @@ void PalGetModuleBounds(HANDLE hOsHandle, _Out_ uint8_t ** ppLowerBound, _Out_ u
+        ULONG               m_cbPdb70;
+    };
+
+    // Grab module bounds so we can do some rough sanity checking before we follow any
+    // RVAs
+    uint8_t * pbModuleLowerBound = NULL;
+    uint8_t * pbModuleUpperBound = NULL;
+    PalGetModuleBounds(hOsHandle, &pbModuleLowerBound, &pbModuleUpperBound);
+
```
```suggestion
```

jkotas commented on src/coreclr/nativeaot/Runtime/windows/PalCommon.cpp:
```diff
@@ -34,6 +34,184 @@ void PalGetModuleBounds(HANDLE hOsHandle, _Out_ uint8_t ** ppLowerBound, _Out_ u
+    PalGetModuleBounds(hOsHandle, &pbModuleLowerBound, &pbModuleUpperBound);
+
+    // Iterate through all debug directory entries. The convention is that debuggers &
+    // profilers typically just use the very last IMAGE_DEBUG_TYPE_CODEVIEW entry.  Treat raw
+    // bytes we read as untrusted.
```
```suggestion
    // profilers typically just use the very last IMAGE_DEBUG_TYPE_CODEVIEW entry.
```
It does not make sense to not trust our own binary.

jkotas commented on src/coreclr/nativeaot/Runtime/windows/PalCommon.cpp:
```diff
@@ -34,6 +34,184 @@ void PalGetModuleBounds(HANDLE hOsHandle, _Out_ uint8_t ** ppLowerBound, _Out_ u
+
+    IMAGE_DATA_DIRECTORY const * pDebugDataDirectory = &rgDataDirectory[IMAGE_DIRECTORY_ENTRY_DEBUG];
+
+    // In Redhawk, modules are loaded as MAPPED, so we don't have to worry about dealing
+    // with FLAT files (with padding missing), so header addresses can be used as is
```
```suggestion
    // We don't have to worry about dealing with FLAT files (with padding missing), so header addresses
    // can be used as is
```

MichalStrehovsky committed:
Apply suggestions from code review

Co-authored-by: Jan Kotas <jkotas@microsoft.com>

MichalStrehovsky committed:
Build break

MichalStrehovsky committed:
Copilot commit

Prompt: The use of mbstowcs_s is incorrect here. The input buffer is UTF-8, output is UTF-16. Replace the incorrect use of mbstowcs_s with MultiByteToWideChar.

MichalStrehovsky commented on src/coreclr/nativeaot/Runtime/windows/PalCommon.cpp:
```diff
@@ -34,6 +34,184 @@ void PalGetModuleBounds(HANDLE hOsHandle, _Out_ uint8_t ** ppLowerBound, _Out_ u
+        // Convert build path from ANSI to UNICODE
+        errno_t ret;
+        size_t cchConverted;
+        ret = mbstowcs_s(
```
I honestly didn't even read any of the code in the method because it's a copy paste from .NET Native codebase where it already passed quality gates and was used in production for years :/

MichalStrehovsky commented:
/azp run runtime-nativeaot-outerloop

MichalStrehovsky committed:
Update issues.targets
